walk/walk.go

```diff
@@ -208,23 +214,35 @@ func buildUpdateRelMap(root string, dirs []string) map[string]bool {
 
 // shouldCall returns true if Walk should call the callback in the
 // directory rel.
-func shouldCall(rel string, mode Mode, updateRels map[string]bool) bool {
-	return mode != UpdateDirsMode || updateRels[rel]
+func shouldCall(rel string, mode Mode, updateParent bool, updateRels map[string]bool) bool {
+	if mode != UpdateDirsMode && mode != UpdateSubdirsMode {
```

Comment:
Having trouble following the logic. I think changing it to a `switch` would be clearer.

```
switch mode {
case VisitAllUpdateSubdirsMode, VisitAllUpdateDirsMode:
  return true
case UpdateSubdirsMode:
  return updateParent || updateRels[rel]
default: // UpdateDirsMode
  return updateRels[rel]
}
```